README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental) [![Travis-CI Build Status](http://badges.herokuapp.com/travis/mdsumner/nchelper?branch=master&env=BUILD_NAME=trusty_release&label=linux)](https://travis-ci.org/mdsumner/nchelper) [![Build Status](http://badges.herokuapp.com/travis/mdsumner/nchelper?branch=master&env=BUILD_NAME=osx_release&label=osx)](https://travis-ci.org/mdsumner/nchelper) [![AppVeyor Build Status](https://ci.appveyor.com/api/projects/status/github/mdsumner/nchelper?branch=master&svg=true)](https://ci.appveyor.com/project/mdsumner/nchelper)

nchelper
========

The goal of nchelper is to provide NetCDF variables as R arrays.

Installation
------------

You can install nchelper from GitHub with:

``` r
# install.packages("devtools")
devtools::install_github("mdsumner/nchelper")
```

Example
-------

This is an example that reads an inbuilt data set of sea surface temperature.

First, find the file and open the variable "sst" as object `a`.

``` r
library(nchelper)
f <- system.file("extdata", "avhrr-only-v2.20180126.nc", package = "nchelper")
a <- nchelper(f, "sst")
dim(a)
#> [1] 1440  720    1    1
names(a)
#> [1] "sst"
```

Now use standard indexing idioms to extract data from this 4D array.

``` r
grcol <- grey(seq(0, 1, length.out = 12))
image(a[,,,], col = grcol)
```

<img src="man/figures/README-four-d-1.png" width="100%" />

``` r
image(a[400:800, 100:300,,], col = grcol)
```

<img src="man/figures/README-four-d-2.png" width="100%" />

``` r
plot(a[,360,,] * 0.01, ylim = c(20, 32), type = "l")
```

<img src="man/figures/README-four-d-3.png" width="100%" />

(Even though we asked for the variable "sst", we could have left it empty for a reminder about any available names.)

``` r
a <- nchelper(f)
#> Warning in nchelper(f): no varname specified, returning time
#>  choose varname from time,zlev,lat,lon,sst,anom,err,ice
a[]
#> [1] 14635
names(a)
#> [1] "time"

## so we go for one of the 2D ones (sst, anom, err, ice)
err <- nchelper(f, "err")
image(err[,,,])
```

<img src="man/figures/README-variables-1.png" width="100%" />

This is a cheap and easy way to investigate the contents of a source.

``` r
image(nchelper(f, "anom")[,,,], col = hcl(240, c = 25, l = 10:90))
```

<img src="man/figures/README-unnamed-chunk-1-1.png" width="100%" /> Please note that this project is released with a [Contributor Code of Conduct](CODE_OF_CONDUCT.md). By participating in this project you agree to abide by its terms.


NEWS.md
# nchelper 0.0.5

* Upgrade package to modern practices. 

# nchelper 0.0-1

* Early explorations. 

R/src-nc.R
#' Tbl
#'
#' 
#' See \code{dplyr::\link[dplyr]{tbl}} for details.
#'
#' @name tbl
#' @rdname dplyr-verbs
#' @keywords internal
#' @importFrom dplyr tbl 
#' @export tbl
#' @importFrom dbplyr op_base
#' @importFrom tibble trunc_mat
#' @importFrom rlang .data
#' @export
#' @examples 
#' f <- system.file("extdata", "S2008001.L3m_DAY_CHL_chlor_a_9km.nc", package = "ncmeta")
#' src_nc(f) %>% tbl("lat")
#' src_nc(f) %>% tbl("palette")
#' 
#' f2 <- system.file("extdata", "avhrr-only-v2.20180126.nc", package = "nchelper")
#' tbl(src_nc(f2), "sst")
#' tbl(src_nc(f2), "zlev")
#' tbl(src_nc(f2), "time")
#' tbl(src_nc(f2), "lat")
#' tbl(src_nc(f2), "lon")
tbl.src_nc <- function(src, variable, ...) {
  varcompare <- variable
  grids <- ncmeta::nc_grids(src$ncsource)
  grid_vars <- dplyr::inner_join(dplyr::filter(grids, variable == varcompare) %>% 
                               dplyr::select(.data$grid), grids, "grid")
  dim_vars <- ncmeta::nc_axes(src$ncsource) %>% dplyr::filter(.data$variable == varcompare) %>% 
    dplyr::inner_join(ncmeta::nc_dims(src$ncsource), c("dimension" = "id"))
  all_vars <- c(dim_vars$name, grid_vars$variable)
  print(all_vars)
  out <- dplyr::make_tbl("lazy", ops = op_base(NULL, unique(all_vars), class= "ncdb"), src = src)
  class(out) <- c("tbl_ncdb", class(out))
  out
}
#' @export
as.data.frame.tbl_ncdb <- function(x, row.names = NULL, optional = NULL,
                                   ..., n = Inf) {
  collect(x, ..., n = n)
}
#' Collect
#'
#' 
#' See \code{utils::\link[utils]{head}} for details.
#'
#' @name head
#' @keywords internal
#' @export
#' @importFrom utils head
head.tbl_ncdb <- function(x, n = 6L, ...) {
  collect(x, ..., n = n)
}


#' Collect
#'
#' 
#' See \code{dplyr::\link[dplyr]{collect}} for details.
#'
#' @name collect
#' @rdname dplyr-verbs
#' @keywords internal
#' @export
#' @importFrom dplyr collect
#' @export collect
collect.tbl_ncdb <- function(x, ...,  n = Inf ) {
  read_nr_ncdb(x, nmax = n)
}

var_dim_get <- function(con, x) {
  tib <- tibble::tibble(variable = x$ops$vars)
  vars <- ncmeta::nc_vars(con) %>% dplyr::inner_join(tib, c("name" = "variable"))
  vars$dim_var <- vars$ndims == max(vars$ndims)
  vars
}
read_nr_ncdb <- function(x, nmax = -1, ...) {
  if (nmax < 1) nmax <- 1
#  if (nmax  < 1) return(tibble::
  nc_con <- RNetCDF::open.nc(x$src$ncsource)
  variables <- var_dim_get(nc_con, x)
  var_variables <- variables$name[variables$dim_var]
  dim_variables <- variables$name[!variables$dim_var]
  #browser()
  dims <- ncmeta::nc_axes(nc_con, var_variables[1L]) %>% dplyr::inner_join(ncmeta::nc_dims(nc_con), c("dimension" = "id"))
  starts <- rep(1L, nrow(dims))
  counts <- rep(1L, nrow(dims))
  counts[1L] <- min(c(nmax, as.integer(prod(dims$length))))
  df <- tibble::as_tibble(stats::setNames(purrr::map(var_variables, ~as.vector(RNetCDF::var.get.nc(nc_con, .x,
                                                       start = starts, count = counts))), 
           var_variables))
  ## TODO expand out the dim vars
  df[dim_variables] <- NA  
  df
}
#' @export
#' @importFrom dbplyr op_vars
dim.tbl_ncdb <- function(x) {
  c(NA, length(op_vars(x$ops)))
}
#' @importFrom dplyr src_tbls
src_tbls.src_nc <- function(x) {
  ncmeta::nc_vars(x$ncsource)$name
}
#' @export
print.src_nc <- function(x, ...) {
  ## needs to be format.src_ncd

  print(sprintf("src: %s", x$name))
  print(sprintf("tbls: %s", paste0(src_tbls(x), collapse = ", ")))
  print(sprintf("origin: %s", dirname(x$ncsource)))
}
#' NetCDF virtual tables
#'
#' @param ncsource file or online link NetCDF source
#'
#' @return a 'src_nc', a virtual-database for a NetCDF source
#' @export
#'
#' @examples
#' f <- system.file("extdata", "S2008001.L3m_DAY_CHL_chlor_a_9km.nc", package = "ncmeta")
#' nc <- src_nc(f)
#' print(nc)
#' \dontrun{
#' nc <- src_nc("http://coastwatch.pfeg.noaa.gov/erddap/griddap/erdQSwind3day")
#' nc
#' }
src_nc <- function(ncsource) {
  structure(list(tbl_f = tbl.src_nc,
                 name = sprintf("<NetCDF: %s>", basename(ncsource)),
                 ncsource = ncsource),
            class = c("src_nc", "src")
  )
}



R/nchelper.R
#' nchelper
#' 
#' Helper functions to treat a NetCDF variable as an R array
#'
#' Create an object to represent a NetCDF variable, primarily to access slices from it with R's Extract syntax for arrays.
#' @param file NetCDF file
#' @param varname NetCDF variable to access
#' @return nchelper object
#' @export
#' @importFrom RNetCDF open.nc file.inq.nc var.inq.nc dim.inq.nc var.get.nc
#' @examples
#' f <- system.file("extdata", "avhrr-only-v2.20180126.nc", package = "nchelper")
#' a <- nchelper(f, "sst")
#' plot(a[,360,,] * 0.01, ylim = c(20, 32), type = "l")
nchelper <- function(file, varname = NULL) {
    #if (!file.exists(file)) stop("File does not exist: ", file)

    nc <- try(RNetCDF::open.nc(file))
    on.exit(RNetCDF::close.nc(nc), add = TRUE)
    if (inherits(nc, "try-error")) stop("Cannot open source: ", file, nc)
    ##on.exit(close.nc(nc))

    fileinq <- file.inq.nc(nc)
    nvars <- fileinq$nvars
    if (is.null(varname)) {
        vars <- character(nvars)
        for (i in seq_len(nvars)) {
            vars[i] <- var.inq.nc(nc, i-1)$name
        }
        varname <- vars[1]
        warning("no varname specified, returning ", varname, "\n choose varname from ", paste(vars, collapse = ","))
    }
    varinq <- var.inq.nc(nc, varname)
    dims <- integer(length(varinq$dimids))
    for (i in seq_along(dims)) dims[i] <- dim.inq.nc(nc, varinq$dimids[i])$length
    structure(.Data = list(file = file, varinq = varinq, dims = dims), class = "nchelper")
}

#' dim
#' 
#' @param x `nchelper` object
#' @export
dim.nchelper <- function(x) {
    x$dims
}
#' names
#' @param x `nchelper`` object
#' @export
names.nchelper <- function(x) {
    x$varinq$name
}
#' Extract
#' 
#' @param x `nchelper` object
#' @param ... index arguments, see `base::Extract`
#' @param drop remove degenerate singleton dimensions, `TRUE` by default
#' @export
"[.nchelper" <- function(x, ..., drop = TRUE) {
    ncdims <- dim(x)
    ndims <- length(ncdims)
    indexlist <- match.call(expand.dots = FALSE)[["..."]]

    ## process indexes
    indexcount <- length(indexlist)
    if (!length(ncdims) == indexcount) stop("index position and number must match variable", names(x), paste0("[", paste(ncdims,collapse = ","), "]"))
    ## TODO: check sensible subsets:
    ## note that we can only handle indexes that are contiguous and increasing
    ## we now have respectable indexes
    starts <- counts <- integer(ndims)
    .processIndex <- function(index, dimsize) {
        if (is.language(index) & index == "") {
            index <- c(1, dimsize)
        } else {
            index <- eval(index)
            index <- c(index[1], length(index))
        }
        index
    }

    ## not monotonically increasing
    ## not within 1:n
    for (i in seq_along(indexlist)) indexlist[[i]] <- .processIndex(indexlist[[i]], ncdims[i])
##return(index)
    nc <- RNetCDF::open.nc(x$file)
    on.exit(RNetCDF::close.nc(nc), add = TRUE)
    d <- var.get.nc(nc, x$varinq$name, start = sapply(indexlist, "[", 1), count = sapply(indexlist, "[", 2))

    d
}


